File: sdk/nodejs/sdk/src/environment/root.ts
import { configFile, environmentFile, identityFile, metadataFile, rootDir } from "./dirs";
import fs from "node:fs"
import {Configuration, MetadataApi, MetadataApiApiKeys} from "../zrok/api/api"

const V = "v0.4"

export class Metadata {
    V: string;
    RootPath: string;

    constructor(V: string, RootPath: string = "") {
        this.V = V
        this.RootPath = RootPath
    }
}

export class ApiEndpoint {
    endpoint: string
    frm: string

    constructor(endpoint: string, frm: string) {
        this.endpoint = endpoint
        this.frm = frm
    }
}

export class Config {
    ApiEndpoint: string

    constructor(ApiEndpoint: string) {
        this.ApiEndpoint = ApiEndpoint
    }
}

export class Environment {
    Token: string
    ZitiIdentity: string
    ApiEndpoint: string

    constructor(Token: string, ZitiIdentity: string, ApiEndpoint: string) {
        this.Token = Token
        this.ZitiIdentity = ZitiIdentity
        this.ApiEndpoint = ApiEndpoint
    }
}

export class Root {
    meta: Metadata
    cfg: Config
    env: Environment

    constructor(meta: Metadata = new Metadata(V, rootDir()), cfg: Config, env: Environment) {
        this.meta = meta
        this.cfg = cfg
        this.env = env
    }

    public HasConfig(): boolean {
        return this.cfg !== undefined && Object.keys(this.cfg).length === 0
    }

    public async Client(): Promise<Configuration> {
        let apiEndpoint = this.ApiEndpoint()
       let mapi = new MetadataApi({basePath: apiEndpoint.endpoint + "/api/v1"})
        mapi.setApiKey(MetadataApiApiKeys, this.env.Token);
       let ver: Promise<string> = mapi.version()

       const regex : RegExp = new RegExp("^(refs/(heads|tags)/)?" + V);
       await ver.then(v => {
        if(!regex.test(v)) {
            throw new Error("Expected a '" + V + "' version, received: '" + v+ "'")
        }
       })

       return conf
    }

    public ApiEndpoint(): ApiEndpoint {
        let apiEndpoint = "https://api-v1.zrok.io"
        let frm = "binary"

        if (this.cfg.ApiEndpoint != "") {
            apiEndpoint = this.cfg.ApiEndpoint
            frm = "config"
        }

        let env = process.env.ZROK_API_ENDPOINT
        if (env != null) {
            apiEndpoint = env
            frm = "ZROK_API_ENDPOINT"
        }

        if (this.IsEnabled()) {
            apiEndpoint = this.env.ApiEndpoint
            frm = "env"
        }

        return new ApiEndpoint(apiEndpoint.replace(/\/+$/, ""), frm)
    }

    public IsEnabled(): boolean {
        return this.env !== undefined && Object.keys(this.env).length > 0
    }

    private PublicIdentityName(): string {
        return "public"
    }

    public EnvironmentIdentityName(): string {
        return "environment"
    }

    public ZitiIdentityNamed(name: string): string {
        return identityFile(name)
    }
}

export function Assert(): boolean {
    if (rootExists()) {
        let meta = loadMetadata()
        return meta.V == V
    }
    return false
}

export function Load(): Root {
    if (rootExists()) {
        return new Root(loadMetadata(), loadConfig(), loadEnvironment())
    }
    throw new Error("unable to load root. Does not exist")
}

function rootExists(): boolean {
    return fs.existsSync(metadataFile())
}

function loadMetadata(): Metadata {
    let mf = metadataFile()
    let data = fs.readFileSync(mf)
    let serial = JSON.parse(data.toString())
    return new Metadata(serial.v)
}

function loadConfig(): Config {
    let cf = configFile()
    if (fs.existsSync(cf)) {    // the config.json file may not be present
        let data = fs.readFileSync(cf)
        let serial = JSON.parse(data.toString())
        return new Config(serial.api_endpoint)    
    } else {
        return new Config('')    
    }
}

function loadEnvironment(): Environment {
    let ef = environmentFile()
    let data = fs.readFileSync(ef)
    let serial = JSON.parse(data.toString())
    return new Environment(serial.zrok_token, serial.ziti_identity, serial.api_endpoint)
}

File: sdk/python/src/zrok/environment/root.py
from dataclasses import dataclass, field
from typing import NamedTuple
from .dirs import identityFile, rootDir, configFile, environmentFile, metadataFile
import os
import json
import zrok_api as zrok
from zrok_api.configuration import Configuration
from zrok_api.models.client_version_check_request import ClientVersionCheckRequest

V = "v1.0"


@dataclass
class Metadata:
    V: str = ""
    RootPath: str = ""


@dataclass
class Config:
    ApiEndpoint: str = ""
    DefaultFrontend: str = ""


@dataclass
class Environment:
    Token: str = ""
    ZitiIdentity: str = ""
    ApiEndpoint: str = ""


class ApiEndpoint(NamedTuple):
    endpoint: str
    frm: str


@dataclass
class Root:
    meta: Metadata = field(default_factory=Metadata)
    cfg: Config = field(default_factory=Config)
    env: Environment = field(default_factory=Environment)

    def HasConfig(self) -> bool:
        return self.cfg != Config()

    def Client(self) -> zrok.ApiClient:
        apiEndpoint = self.ApiEndpoint()

        cfg = Configuration()
        cfg.host = apiEndpoint[0] + "/api/v2"

        # Update: Configure authentication token
        # The token needs to be set with 'key' instead of 'x-token'
        # This matches the securityDefinitions in the OpenAPI spec
        cfg.api_key["key"] = self.env.Token

        # Create the API client with the configured authentication
        auth_client = zrok.ApiClient(configuration=cfg)
        self.client_version_check(auth_client)

        return auth_client

    def ApiEndpoint(self) -> ApiEndpoint:
        apiEndpoint = "https://api-v2.zrok.io"
        frm = "binary"

        if self.cfg.ApiEndpoint != "":
            apiEndpoint = self.cfg.ApiEndpoint
            frm = "config"

        env = os.getenv("ZROK_API_ENDPOINT")
        if env != "":
            apiEndpoint = env
            frm = "ZROK_API_ENDPOINT"

        if self.IsEnabled():
            apiEndpoint = self.env.ApiEndpoint
            frm = "env"

        return ApiEndpoint(apiEndpoint.rstrip("/"), frm)

    def IsEnabled(self) -> bool:
        return self.env != Environment()

    def PublicIdentityName(self) -> str:
        return "public"

    def EnvironmentIdentityName(self) -> str:
        return "environment"

    def ZitiIdentityNamed(self, name: str) -> str:
        return identityFile(name)

    def client_version_check(self, zrock_client):
        """Check if the client version is compatible with the API."""
        metadata_api = zrok.MetadataApi(zrock_client)
        try:
            # Perform version check using the client_version_check method
            request = ClientVersionCheckRequest(client_version=V)
            response = metadata_api.client_version_check_with_http_info(
                body=request,
            )

            # Check if the response status code is 200 OK
            if response.status_code != 200:
                raise Exception(f"Client version check failed: Unexpected status code {response.status_code}")

            # Success case - status code is 200 and empty response body is expected
            return

        except Exception as e:
            raise Exception(f"Client version check failed: {str(e)}")


def Default() -> Root:
    r = Root()
    root = rootDir()
    r.meta = Metadata(V=V, RootPath=root)
    return r


def Assert() -> bool:
    exists = __rootExists()
    if exists:
        meta = __loadMetadata()
        return meta.V == V
    return False


def Load() -> Root:
    r = Root()
    if __rootExists():
        r.meta = __loadMetadata()
        r.cfg = __loadConfig()
        r.env = __loadEnvironment()
    else:
        r = Default()
    return r


def __rootExists() -> bool:
    mf = metadataFile()
    return os.path.isfile(mf)


def __assertMetadata():
    pass


def __loadMetadata() -> Metadata:
    mf = metadataFile()
    with open(mf) as f:
        data = json.load(f)
        return Metadata(V=data["v"])


def __loadConfig() -> Config:
    cf = configFile()
    try:
        with open(cf) as f:
            data = json.load(f)
            return Config(
                ApiEndpoint=data.get("api_endpoint", ""),
                DefaultFrontend=data.get("default_frontend", "")
            )
    except (FileNotFoundError, json.JSONDecodeError):
        return Config(ApiEndpoint="", DefaultFrontend="")


def isEnabled() -> bool:
    ef = environmentFile()
    return os.path.isfile(ef)


def __loadEnvironment() -> Environment:
    ef = environmentFile()
    with open(ef) as f:
        data = json.load(f)
        return Environment(
            Token=data["zrok_token"],
            ZitiIdentity=data["ziti_identity"],
            ApiEndpoint=data["api_endpoint"])
